rust/bridge/node/src/lib.rs

```diff
@@ -3,13 +3,289 @@
+                                (*handle)
+                                    .to_string(cx)
+                                    .map_err(|e| e.to_string())?
+                                    .value(cx)
+                            );
+                            Ok(None) // fixme!
+                        }
+                        Err(_) => {
+                            eprintln!("not an object?");
+                            if value.is_a::<JsNull, _>(cx) {
+                                eprintln!("you returned null");
+                                Ok(None)
+                            } else {
+                                eprintln!("not an object");
+                                Err("result must be an object".into())
+                            }
+                        }
+                    }
+                }
+                Err(error) => {
+                    eprintln!("downcast failed :/");
+                    Err(error
+                        .to_string(cx)
+                        .expect("can convert to string")
+                        .value(cx))
+                }
+            }
+        })
+        .await
+    }
+
+    async fn do_save_sender_key(
+        &self,
+        name: SenderKeyName,
+        record: SenderKeyRecord,
+    ) -> Result<(), String> {
+        let store_object_shared = self.store_object.clone();
+        JsFuture::get_promise(&self.js_queue, move |cx| {
+            let store_object = store_object_shared.to_inner(cx);
+            let name: Handle<JsValue> = boxed_object(cx, name)?;
+            let record: Handle<JsValue> = boxed_object(cx, record)?;
+            let result = call_method(
+                cx,
+                store_object,
+                "_saveSenderKey",
+                std::iter::once(name).chain(std::iter::once(record)),
+            )?
+            .downcast_or_throw(cx)?;
+            store_object_shared.finalize(cx);
+            Ok(result)
+        })
+        .then(|cx, result| match result {
+            Ok(value) => match value.downcast::<JsUndefined, _>(cx) {
+                Ok(_) => Ok(()),
+                Err(_) => Err("name must be a string".into()),
+            },
+            Err(error) => Err(error
+                .to_string(cx)
+                .expect("can convert to string")
+                .value(cx)),
+        })
+        .await
+    }
+}
+
+impl<'a> Finalize for NodeSenderKeyStore<'a> {
+    fn finalize<'b, C: neon::prelude::Context<'b>>(self, cx: &mut C) {
+        self.store_object.finalize(cx)
+    }
+}
+
+#[async_trait(?Send)]
+impl<'a> SenderKeyStore for NodeSenderKeyStore<'a> {
+    async fn load_sender_key(
+        &mut self,
+        sender_key_name: &SenderKeyName,
+        _ctx: libsignal_protocol::Context,
+    ) -> Result<Option<SenderKeyRecord>, SignalProtocolError> {
+        eprintln!("load_sender_key({:?})", sender_key_name);
+        // XXX clone()
+        self.do_get_sender_key(sender_key_name.clone())
```

Comment:
…but we shouldn't need to clone *twice.* (I imagine this was moving things around.)